src/provisioning/Provisioner.ts

```diff
@@ -0,0 +1,568 @@
+    const userId = req.userId;
+    if (!userId) {
+        throw new ApiError("Missing user ID", ErrCode.BadValue);
+    }
+    return userId;
+};
+
+type RequestHandler = (req: ProvisioningRequest, res: Response, next?: NextFunction) => Promise<unknown>;
+
+export class Provisioner extends ProvisioningApi {
+    constructor(
+        readonly main: Main,
+        readonly appService: AppService,
+        readonly config: StrictProvisionerConfig,
+    ) {
+        super(
+            main.datastore,
+            {
+                provisioningToken: config.secret,
+                apiPrefix: "/_matrix/provision",
+                ratelimit: (config.ratelimit ?? true),
+                disallowedIpRanges: config.open_id_disallowed_ip_ranges,
+                openIdOverride: config.open_id_overrides
+                    ? Object.fromEntries(Object.entries(config.open_id_overrides)
+                        .map(([id, url]) => [id, new URL(url)])
+                    )
+                    : undefined,
+                widgetTokenPrefix: "slackbr-wdt-",
+                widgetFrontendLocation: config.widget ? "public" : undefined,
+                // Use the bridge express application unless a config was specified for provisioning
+                expressApp: config.http ? undefined : appService.expressApp,
+            },
+        );
+
+        if (!config.enabled) {
+            log.info("Provisioning disabled, endpoints will respond with an error code");
+            // Disable all provision endpoints, responding with an error instead
+            this.baseRoute.use((req, res, next) => next(new ApiError("Provisioning not enabled", ErrCode.DisabledFeature)));
+            return;
+        }
+
+        if (this.store instanceof NedbDatastore) {
+            log.warn(
+                "Provisioner is incompatible with NeDB store. Widget requests will not be handled."
+            );
+        }
+
+        const wrapHandler = (handler: RequestHandler) =>
+            async (req: ProvisioningRequest, res: Response) => {
+                // Wrap handlers so that they can follow the common style of `return res.json(...)`
+                // This is only necessary because ProvisioningApi requires return of Promise<void>
+                try {
+                    await handler.call(this, req, res);
+                } catch (e) {
+                    if (res.headersSent) {
+                        return;
+                    }
+                    if (e instanceof SlackProvisioningError || e instanceof ApiError) {
+                        e.apply(res);
+                    } else {
+                        req.log.error("Unknown error:", e);
+                        // Send a generic error
+                        const err = new ApiError("An internal error occurred");
+                        err.apply(res);
+                    }
+                }
+            };
+
+        this.addRoute("post", "/getbotid", wrapHandler(this.getBotId), "getBotId");
```

Comment:
Can we rename this endpoint to /getbotinfo (with perhaps a `/getbotid` fallback). The naming is terrible.